File: src/main/java/com/refinedmods/refinedstorage/tile/data/TileDataManager.java
package com.refinedmods.refinedstorage.tile.data;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.network.tiledata.TileDataParameterUpdateMessage;
import net.minecraft.world.level.block.entity.BlockEntity;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;

public class TileDataManager {
    private static final Map<Integer, TileDataParameter> REGISTRY = new HashMap<>();
    private static int lastId = 0;
    private final BlockEntity tile;

    private final List<TileDataParameter> parameters = new ArrayList<>();
    private final List<TileDataParameter> watchedParameters = new ArrayList<>();

    private final List<TileDataWatcher> watchers = new CopyOnWriteArrayList<>();

    public TileDataManager(BlockEntity tile) {
        this.tile = tile;
    }

    public static void registerParameter(TileDataParameter parameter) {
        parameter.setId(lastId);

        REGISTRY.put(lastId++, parameter);
    }

    public static TileDataParameter getParameter(int id) {
        return REGISTRY.get(id);
    }

    public static void setParameter(TileDataParameter parameter, Object value) {
        RS.NETWORK_HANDLER.sendToServer(new TileDataParameterUpdateMessage(parameter, value));
    }

    public BlockEntity getTile() {
        return tile;
    }

    public void addParameter(TileDataParameter parameter) {
        parameters.add(parameter);
    }

    public List<TileDataParameter> getParameters() {
        return parameters;
    }

    public void addWatchedParameter(TileDataParameter parameter) {
        addParameter(parameter);

        watchedParameters.add(parameter);
    }

    public List<TileDataParameter> getWatchedParameters() {
        return watchedParameters;
    }

    public void addWatcher(TileDataWatcher listener) {
        watchers.add(listener);
    }

    public void removeWatcher(TileDataWatcher listener) {
        watchers.remove(listener);
    }

    public void sendParameterToWatchers(TileDataParameter parameter) {
        watchers.forEach(l -> l.sendParameter(false, parameter));
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/tile/data/TileDataManager.kt
package com.refinedmods.refinedstorage.tile.data

import com.refinedmods.refinedstorage.RS
import com.refinedmods.refinedstorage.network.tiledata.TileDataParameterUpdateMessage
import net.minecraft.block.entity.BlockEntity
import java.lang.ClassCastException
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList
import java.util.function.Consumer

class TileDataManager(
        val tile: BlockEntity
) {
    private val parameters: MutableList<TileDataParameter<*, *>> = ArrayList()
    private val watchedParameters: MutableList<TileDataParameter<*, *>> = ArrayList()
    private val watchers: MutableList<TileDataWatcher> = CopyOnWriteArrayList()

    fun addParameter(parameter: TileDataParameter<*, *>) {
        parameters.add(parameter)
    }

    fun getParameters(): List<TileDataParameter<*, *>> {
        return parameters
    }

    fun addWatchedParameter(parameter: TileDataParameter<*, *>) {
        addParameter(parameter)
        watchedParameters.add(parameter)
    }

    fun getWatchedParameters(): List<TileDataParameter<*, *>> {
        return watchedParameters
    }

    fun addWatcher(listener: TileDataWatcher) {
        watchers.add(listener)
    }

    fun removeWatcher(listener: TileDataWatcher?) {
        watchers.remove(listener)
    }

    fun sendParameterToWatchers(parameter: TileDataParameter<*, *>?) {
        // TODO Send params
//        watchers.forEach(Consumer { l: TileDataWatcher -> l.sendParameter(false, parameter) })
    }

    companion object {
        private var LAST_ID = 0
        private val REGISTRY: MutableMap<Int, TileDataParameter<Any?, BlockEntity?>> = HashMap()
        fun registerParameter(parameter: TileDataParameter<Any?, BlockEntity?>) {
            parameter.id = LAST_ID
            REGISTRY[LAST_ID++] = parameter
        }

        fun <T: Any?, E: BlockEntity?> getParameter(id: Int): TileDataParameter<T?, E?>? {
            return try {
                REGISTRY[id] as TileDataParameter<T?, E?>?
            } catch (e: ClassCastException){
                null
            }
        }

        fun setParameter(parameter: TileDataParameter<*, *>?, value: Any?) {
            // TODO Setup network handler
            //RS.NETWORK_HANDLER.sendToServer(TileDataParameterUpdateMessage(parameter, value))
        }
    }

}